pkg/mcs/resource_manager/server/server.go

```diff
@@ -0,0 +1,305 @@
+		return err
+	}
+	if err := s.startServer(); err != nil {
+		return err
+	}
+	return s.mux.Serve()
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+
+	s.grpcServer.Stop()
+	s.httpServer.Shutdown(s.ctx)
+	s.serverLoopWg.Wait()
+
+	if s.etcdClient != nil {
+		if err := s.etcdClient.Close(); err != nil {
+			log.Error("close etcd client meet error", errs.ZapError(errs.ErrCloseEtcdClient, err))
+		}
+	}
+
+	if s.httpClient != nil {
+		s.httpClient.CloseIdleConnections()
+	}
+}
+
+// GetClient returns builtin etcd client.
+func (s *Server) GetClient() *clientv3.Client {
+	return s.etcdClient
+}
+
+// GetHTTPClient returns builtin http client.
+func (s *Server) GetHTTPClient() *http.Client {
+	return s.httpClient
+}
+
+// AddStartCallback adds a callback in the startServer phase.
+func (s *Server) AddStartCallback(callbacks ...func()) {
+	s.startCallbacks = append(s.startCallbacks, callbacks...)
+}
+
+// IsPrimary returns whether the server is primary, if there is etcd server in the server, it will return whether it is leader.
+func (s *Server) IsPrimary() bool {
+	// TODO: implement this
+	return true
+}
+
+// AddPrimaryCallback adds a callback when the server becomes primary, if there is etcd server in the server, it means the server becomes leader.
+func (s *Server) AddPrimaryCallback(callbacks ...func(context.Context)) {
+	s.primaryCallbacks = append(s.primaryCallbacks, callbacks...)
+}
+
+func (s *Server) initClient() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	endpoint, err := url.Parse(s.cfg.BackendEndpoints)
+	if err != nil {
+		return err
+	}
+	s.etcdClient, s.httpClient, err = etcdutil.CreateClients(tlsConfig, []url.URL{*endpoint})
+	return err
+}
+
+func (s *Server) startServer() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	if tlsConfig != nil {
+		l, err := tls.Listen("tcp", s.cfg.ListenAddr, tlsConfig)
+		if err != nil {
+			return err
+		}
+		s.mux = cmux.New(l)
+	} else {
+		l, err := net.Listen("tcp", s.cfg.ListenAddr)
+		if err != nil {
+			return err
+		}
+		s.mux = cmux.New(l)
+	}
+
+	grpcL := s.mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
+	httpL := s.mux.Match(cmux.Any())
+
+	manager := NewManager(s)
+	s.service = &Service{
+		ctx:     s.ctx,
+		manager: manager,
+	}
+
+	s.grpcServer = grpc.NewServer()
+	s.service.RegisterGRPCService(s.grpcServer)
+
+	handler, _ := SetUpRestHandler(s.service)
+	s.httpServer = &http.Server{
+		Handler:           handler,
+		ReadTimeout:       5 * time.Minute,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	go func() {
+		defer s.serverLoopWg.Done()
+		s.grpcServer.Serve(grpcL)
+	}()
+	go func() {
+		defer s.serverLoopWg.Done()
+		s.httpServer.Serve(httpL)
+	}()
+	s.serverLoopWg.Add(2)
```

Comment:
Shouldn't move this line before the two go func() above? Otherwise there is race condition when a go routine exit abnormally before it reaches this line.